sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs

```diff
@@ -44,6 +43,59 @@ public virtual bool IsClosed
             private set => _closed = value;
         }
 
+        /// <summary>
+        /// An event that can be subscribed to for notification when the client connects to the service.
+        /// </summary>
+#pragma warning disable AZC0002
+#pragma warning disable AZC0003
+        public event Func<ServiceBusConnectionEventArgs, Task> ConnectedAsync;
+#pragma warning restore AZC0003
+#pragma warning restore AZC0002
+
+        /// <summary>
+        /// An event that can be subscribed to for notification when the client disconnects from the service.
+        /// No action is required when the client temporarily disconnects due to a transient network or service issue, as the client will attempt
+        /// to re-establish the connection automatically on the next operation. If <see cref="DisposeAsync"/> is
+        /// called, then the connection will not be re-established.
+        /// </summary>
+#pragma warning disable AZC0003
+#pragma warning disable AZC0002
+        public event Func<ServiceBusConnectionEventArgs, Task> DisconnectedAsync;
+#pragma warning restore AZC0002
+#pragma warning restore AZC0003
+
+        /// <summary>
+        /// Gets whether or not the client is currently connected to the service. No action is required when the client temporarily
+        /// disconnects due to a transient network or service issue, as the client will attempt to re-establish the connection
+        /// automatically on the next operation. As such, it is not recommended that this property is checked before performing operations with the <see cref="ServiceBusClient"/>
+        /// or any of the other Service Bus types. This property should be used only for diagnostic information.
+        /// </summary>
+        public virtual bool IsConnected => Connection.IsConnected;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <returns></returns>
+        protected internal virtual async Task OnConnectedAsync(ServiceBusConnectionEventArgs args)
+        {
+            if (ConnectedAsync != null)
+            {
+                await ConnectedAsync(args).ConfigureAwait(false);
+            }
```

Comment:
```suggestion
            var handler = ConnectedAsync;
           
            if (handler != null)
            {
                await handler(args).ConfigureAwait(false);
            }
```
This is a race condition, since we're not blocking handlers being added/removed at any time.